src/contract_clients/utils.rs

```diff
@@ -110,125 +115,152 @@ pub(crate) enum DeclarationInput<'a> {
     // inputs : sierra_path, casm_path
     DeclarationInputs(String, String, RpcAccount<'a>),
     // input : artifact_path
-    LegacyDeclarationInputs(String, String),
+    LegacyDeclarationInputs(String, String, &'a JsonRpcClient<HttpTransport>),
 }
 
 #[allow(private_interfaces)]
-pub async fn declare_contract(input: DeclarationInput<'_>) -> FieldElement {
+pub async fn declare_contract(input: DeclarationInput<'_>) -> Felt {
     match input {
         DeclarationInputs(sierra_path, casm_path, account) => {
-            let (class_hash, sierra) = account.declare_contract_params_sierra(&sierra_path, &casm_path);
+            let contract_artifact: SierraClass =
+                serde_json::from_reader(std::fs::File::open(sierra_path).unwrap()).unwrap();
+
+            let contract_artifact_casm: CompiledClass =
+                serde_json::from_reader(std::fs::File::open(casm_path).unwrap()).unwrap();
+            let class_hash = contract_artifact_casm.class_hash().unwrap();
+            let sierra_class_hash = contract_artifact.class_hash().unwrap();
+
+            if account.provider().get_class(BlockId::Tag(Pending), sierra_class_hash).await.is_ok() {
+                return sierra_class_hash;
+            }
+
+            let flattened_class = contract_artifact.flatten().unwrap();
 
             account
-                .declare(Arc::new(sierra.clone()), class_hash)
+                .declare_v3(Arc::new(flattened_class), class_hash)
+                .gas(0)
                 .send()
                 .await
                 .expect("Error in declaring the contract using Cairo 1 declaration using the provided account");
-            sierra.class_hash()
+            sierra_class_hash
         }
-        LegacyDeclarationInputs(artifact_path, url) => {
+        LegacyDeclarationInputs(artifact_path, url, provider) => {
             let contract_abi_artifact_temp: LegacyContractClass = serde_json::from_reader(
```

Comment:
```suggestion
            let contract_abi_artifact: LegacyContractClass = serde_json::from_reader(
```
we can just call it `contract_abi_artifact` and override later (this was already there before but i think it's a small change so we can maybe do it now 🙂)

Reply:
resolved